contracts/service-provider-directory/src/state/services.rs

```diff
@@ -0,0 +1,293 @@
+use cosmwasm_std::{Addr, Order, StdError, StdResult, Storage};
+use cw_storage_plus::{Bound, Index, IndexList, IndexedMap, MultiIndex};
+use nym_service_provider_directory_common::{NymAddress, Service, ServiceId};
+
+use crate::{
+    constants::{
+        MAX_NUMBER_OF_ALIASES_FOR_NYM_ADDRESS, MAX_NUMBER_OF_PROVIDERS_PER_OWNER,
+        SERVICE_DEFAULT_RETRIEVAL_LIMIT, SERVICE_MAX_RETRIEVAL_LIMIT,
+    },
+    error::{ContractError, Result},
+};
+
+const SERVICES_PK_NAMESPACE: &str = "sernames";
+const SERVICES_OWNER_IDX_NAMESPACE: &str = "serown";
+const SERVICES_NYM_ADDRESS_IDX_NAMESPACE: &str = "sernyma";
+
+struct ServiceIndex<'a> {
+    pub(crate) nym_address: MultiIndex<'a, String, Service, ServiceId>,
+    pub(crate) owner: MultiIndex<'a, Addr, Service, ServiceId>,
+}
+
+impl<'a> IndexList<Service> for ServiceIndex<'a> {
+    fn get_indexes(&'_ self) -> Box<dyn Iterator<Item = &'_ dyn Index<Service>> + '_> {
+        let v: Vec<&dyn Index<Service>> = vec![&self.nym_address, &self.owner];
+        Box::new(v.into_iter())
+    }
+}
+
+fn services<'a>() -> IndexedMap<'a, ServiceId, Service, ServiceIndex<'a>> {
+    let indexes = ServiceIndex {
+        nym_address: MultiIndex::new(
```

Comment:
Should a `nym_address` be a `MultiIndex` or maybe a `UniqueIndex`? Do we want to allow registering multiple different service providers with the same `nym_address`?

Reply:
The same thought occurred to me as well, but after pondering it a bit, yeah I think it might be useful to register the same nym-address under multiple names

Reply:
but isn't it the case that the only things announced are the service type, owner and the nym address? So in that case would that duplicate nym address be associated with multiple owners?

Reply:
Yep correct. Even without allowing multiple names it's still the case that someone might announce a service they don't own themselves ...